docs/core/build/distribution-packaging.md

```diff
@@ -72,48 +91,29 @@ The rest of the version isn't included in the version name. This allows the OS p
-| dotnet-sdk-[major].[minor]              | dotnet-sdk-2.1         | Latest sdk for specific runtime |                    | dotnet-sdk-[major].[minor].[latest sdk feat]xx | \<sdk version>     |
-| dotnet-sdk-[major].[minor].[sdk feat]xx | dotnet-sdk-2.1.3xx     | Specific sdk feature release    | (3),(4)            | aspnetcore-runtime-[major].[minor]             | \<sdk version>     |
-| aspnetcore-runtime-[major].[minor]      | aspnetcore-runtime-2.1 | Specific ASP.NET Core runtime   | (6),[(7)]          | dotnet-runtime-[major].[minor]                 | \<runtime version> |
-| dotnet-runtime-[major].[minor]          | dotnet-runtime-2.1     | Specific runtime                | (5)                | host-fxr:\<runtime version>+                   | \<runtime version> |
-| dotnet-host-fxr                         | dotnet-host-fxr        | _dependency_                    | (2)                | host:\<runtime version>+                       | \<runtime version> |
-| dotnet-host                             | dotnet-host            | _dependency_                    | (1),(8),(9),(10)   |                                                | \<runtime version> |
+| Name                              | Example            | Use case: Install ...           | Contains         | Dependencies                      | Version            |
+|-----------------------------------|--------------------|---------------------------------|------------------|-----------------------------------|--------------------|
+| dotnet-sdk-[major]                | dotnet-sdk-2       | Latest sdk for runtime major    |                  | dotnet-sdk-[major].[latestminor]  | \<sdk version>     |
+| dotnet-sdk-[major].[minor]        | dotnet-sdk-2.1     | Latest sdk for specific runtime | (3),(4),(13)     | aspnet-runtime-[major].[minor], dotnet-netcoreapp-targeting-pack[major].[minor], dotnet-aspnet-targeting-pack[major].[minor], dotnet-netstandard-targeting-pack-[netst_major].[netstd_minor] | \<sdk version>     |
+| aspnet-runtime-[major].[minor]    | aspnet-runtime-2.1 | Specific ASP.NET Core runtime   | (6),[(7)]        | dotnet-runtime-[major].[minor]    | \<runtime version> |
+| dotnet-runtime-[major].[minor]    | dotnet-runtime-2.1 | Specific runtime                | (5)              | host-fxr:\<runtime version>+      | \<runtime version> |
+| dotnet-host-fxr                   | dotnet-host-fxr    | _dependency_                    | (2)              | host:\<runtime version>+          | \<runtime version> |
+| dotnet-host                       | dotnet-host        | _dependency_                    | (1),(8),(9),(10) |                                   | \<runtime version> |
+| dotnet-netcoreapp-targeting-pack-[major].[minor]    |  | _dependency_                    | (12)             |                                   | \<pack version>    |
```

Comment:
This doesn't have `netcoreapp` in the Microsoft build:

```suggestion
| dotnet-targeting-pack-[major].[minor]    |  | _dependency_                    | (12)             |                                   | \<pack version>    |
```

I don't feel strongly either way--I think what matters here (as far as naming matters :smile:) is making these consistent, which you could do any number of ways:

```
dotnet-runtime-[major].[minor]
dotnet-targeting-pack-[major].[minor]
aspnet-runtime-[major].[minor]
aspnet-targeting-pack-[major].[minor]
```

Reply:
I'm not opinionated about the names either, I care about having the same names for Red Hat/Fedora/Microsoft packages.

Reply:
Here are the Preview 6 RPMs I know about from Microsoft build output, for reference:

**ASP.NET Core (AspNetCore)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| **aspnetcore-runtime-3.0** | 3.0.0 | 0.1.preview6.19307.2 | dotnet-runtime-3.0 |
| aspnetcore-targeting-pack-3.0 | 3.0.0 | 0.1.preview6.19307.2 | dotnet-targeting-pack-3.0 |

**SDK (Core-SDK)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| **dotnet-sdk-3.0** | 3.0.100_preview6_012264 | 1 | dotnet-runtime-3.0, dotnet-targeting-pack-3.0, dotnet-apphost-pack-3.0, netstandard-targeting-pack-2.1, aspnetcore-runtime-3.0, aspnetcore-targeting-pack-3.0 |

**Runtime (Core-Setup)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| dotnet-apphost-pack-3.0 | 3.0.0 | 0.1._preview6_27804_01 |  |
| dotnet-host | 3.0.0 | 0.1.preview6_27804_01 |  |
| dotnet-hostfxr-3.0 | 3.0.0 | 0.1.preview6_27804_01 | dotnet-host |
| **dotnet-runtime-3.0** | 3.0.0 | 0.1.preview6_27804_01 | dotnet-hostfxr-3.0, dotnet-runtime-deps-3.0 |
| dotnet-runtime-deps-3.0 | 3.0.0 | 0.1.preview6_27804_01 | `<per-distro>` |
| dotnet-targeting-pack-3.0 | 3.0.0 | 0.1._preview6_27804_01 |  |
| netstandard-targeting-pack-2.1 | 2.1.0 | 0.1._preview6_27804_01 |  |

```sh
rpm -qp --queryformat '| %{NAME} | %{VERSION} | %{RELEASE} | [%{REQUIRES}, ] |\n' *.rpm
# (With some cleanup)
```

---

The versions look odd (probably because we haven't put preview packages on a package repository so they aren't ironed out) but I think the granularity and dependencies are what we're expecting to go with. I bolded the expected "entry point" packages for install commands.

(Going into this, I assumed the existing set was common knowledge and this PR was about making changes for Red Hat and other distros, but I don't actually know how I would have generated those tables without access to internal data. 😕 We need [public BAR access](https://github.com/dotnet/arcade/issues/1818).)

Reply:
> I assumed the existing set was common knowledge and this PR was about making changes for Red Hat and other distros, but I don't actually know how I would have generated those tables without access to internal data.

There has been no public info about this, that's why I've started the PR.

Based on your list, I'll make these changes:
- `aspnetcore` vs `aspnet` -> `aspnetcore`
- remove `dotnet-` prefix
- `dotnet-hostfxr` vs `dotnet-host-fxr`, Microsoft is using the former, this doc (and Fedora/RHEL) the latter. -> changing to `dotnet-hostfxr`.

I think for Fedora/RHEL we'll stick to `dotnet-host-fxr` since we are using that name already.

- `dotnet-apphost-pack-3.0`, I'll add this.

The packs have _full version number folders:
```
├── Microsoft.AspNetCore.App.Ref
│   └── 3.0.0-preview6.19307.2
├── Microsoft.NETCore.App.Host.fedora.30-x64
│   └── 3.0.0-preview6-27804-01
├── Microsoft.NETCore.App.Ref
│   └── 3.0.0-preview6-27804-01
├── Microsoft.WindowsDesktop.App.Ref
│   └── 3.0.0-preview6-27804-01
└── NETStandard.Library.Ref
    └── 2.1.0-preview6-27804-01
```
Does that imply that the sdk supports some sort of _roll forward_ if successive versions of the same package use different version numbers here?
Or will this in be: `[major].[minor].0` for released versions, and not change for successive versions?

Reply:
> The packs have _full version number folders: [...]
>
> Does that imply that the sdk supports some sort of _roll forward_ if successive versions of the same package use different version numbers here?
> Or will this in be: `[major].[minor].0` for released versions, and not change for successive versions?

https://github.com/dotnet/designs/pull/50 touches on the philosophy around this:

> Targeting pack package versions will generally not increase past major.minor.0 where major.minor matches the the corresponding two-part TFM. For example, when there is a 3.0.1 .NET Core runtime, the targeting pack will likely remain at 3.0.0. This follows from the fact that targeting packs represent public API surface, which must not change in a patch version of the runtime. (With that said, we can reserve the right to modify the targeting pack in a patch release to fix a severe bug. In the rare event that this occurs, the targeting pack patch version could be incremented past 0.)
> 
> During the prerelease phase of a new major.minor version of the runtime, the targeting pack will version 1:1 with the runtime. This is necessary as different prerelease versions will have different surface area.

The SDK has the version in `Microsoft.NETCoreSdk.BundledVersions.props` as of preview6 ([generated here](https://github.com/dotnet/core-sdk/blob/master/src/redist/targets/GenerateBundledVersions.targets)):

```xml
    <KnownFrameworkReference Include="Microsoft.NETCore.App"
                              TargetFramework="netcoreapp3.0"
                              RuntimeFrameworkName="Microsoft.NETCore.App"
                              DefaultRuntimeFrameworkVersion="3.0.0-preview6-27804-01"
                              LatestRuntimeFrameworkVersion="3.0.0-preview6-27804-01"
                              TargetingPackName="Microsoft.NETCore.App.Ref"
                              TargetingPackVersion="3.0.0-preview6-27804-01"
                              AppHostPackNamePattern="runtime.**RID**.Microsoft.NETCore.DotNetAppHost"
                              AppHostRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
                              RuntimePackNamePatterns="runtime.**RID**.Microsoft.NETCore.App;runtime.**RID**.Microsoft.NETCore.DotNetHostResolver;runtime.**RID**.Microsoft.NETCore.DotNetHostPolicy"
                              RuntimePackRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
                              PackagesToReference="Microsoft.NETCore.App/3.0.0-preview6-27804-01"
                              IsTrimmable="true"
                              />
```

I imagine there's a project-driven way to use a specific version, but I don't know of any roll forward based on what's on disk.

Reply:
> For example, when there is a 3.0.1 .NET Core runtime, the targeting pack will likely remain at 3.0.0. This follows from the fact that targeting packs represent public API surface, which must not change in a patch version of the runtime.

ok, this matches my expectations.

This was about the version that shows up in the _folders_. Now for the version we put onto the pack _packages_. I think for practical reasons we could use the `<sdk version>`? It seems simpler if successive versions of .NET Core can build a package with a higher version (even if the content is the same)? Or should we for some reason version these based on what is in the package?

Reply:
> successive versions of .NET Core can build a package with a higher version (even if the content is the same)?

This part sounds very similar to the CoreFX platform extension package problem: https://github.com/dotnet/source-build/issues/923. The core is that there's no assurance it's possible to build old assets using a new commit because the Microsoft build process doesn't exercise it.

Independent of that, IMO a reason to not use the SDK version is to let users get lower-versioned targeting packs without having to look up too much about how versioning works. I.e. if they want to target netstandard 2.1, they get `netstandard-targeting-pack-2.1`. They don't have to figure out what SDK version was built with that targeting pack (3.0).

What makes it more practical to use SDK version? Would it help to split up source-build into "runtime" (includes runtime targeting packs), "netstandard", and "sdk" builds? (Or imagine it split up that way?)

Reply:
> the CoreFX platform extension package problem

I don't understand what this is about. How does source-build relate to _specific non-shared-framework packages_?

> IMO a reason to not use the SDK version is to let users get lower-versioned targeting packs without having to look up too much about how versioning works. 

Suppose there is no .NET Core on my machine, and I'd like to install .NET Core 3.1 sdk and 3.0 targeting pack. Then I'd:
```
$ dnf install dotnet-sdk-3.1 dotnet-targeting-pack-3.0
```
The versions the user cares about are part of the package names, the package versions don't show up here.

> What makes it more practical to use SDK version? Would it help to split up source-build into "runtime" (includes runtime targeting packs), "netstandard", and "sdk" builds? (Or imagine it split up that way?)

In once case source-build `3.0.100` and `3.0.200` both build a `dotnet-targeting-pack-3.0` with version `3.0.0` And in the other, the versions are `3.0.100` and `3.0.200`.
I was thinking that for a package maintainer, maybe it is simpler to not have to deal with packages that end up with the same version numbers.
But it seems this is already happening for the sdk/runtime (runtime `2.2.6` is used with sdk `2.2.108`, `2.2.205`, and `2.2.301`).

Reply:
> > the CoreFX platform extension package problem
> 
> I don't understand what this is about. How does source-build relate to _specific non-shared-framework packages_?

I'm not quite sure what you mean by shared. But in any case, my point here was more about the problem of building runtime 3.0.1 (say there were no targeting pack fixes), but trying to build targeting packs that match the 3.0.0 outputs. This is what wouldn't reliably work for the same reasons as https://github.com/dotnet/source-build/issues/923. It seemed like some of this might depend on that working, so I wanted to bring it up, maybe I was mistaken. Either way, that's a big topic that calls for a general solution and doesn't need to be fleshed out here.

> The versions the user cares about are part of the package names, the package versions don't show up here.

Good clarification, I've always assumed the major/minor in the package name and the package version should match as some sort of convention, but I don't know a reason for it. I do agree that keeping the package version consistent across a build of the packages makes sense.

> In once case source-build `3.0.100` and `3.0.200` both build a `dotnet-targeting-pack-3.0` with version `3.0.0` And in the other, the versions are `3.0.100` and `3.0.200`.
> I was thinking that for a package maintainer, maybe it is simpler to not have to deal with packages that end up with the same version numbers.
> But it seems this is already happening for the sdk/runtime (runtime `2.2.6` is used with sdk `2.2.108`, `2.2.205`, and `2.2.301`).

Thanks, I agree. It's somewhat confusing that two builds produce the same packages, but at least with different package versions there's an obvious way to resolve it.

I think splitting up the builds resolves this too. But now I'm thinking it makes sense to document what can be done right now (one package version per build, based on the "root" SDK artifact) and come back to revise if splitting becomes possible.

Reply:
> I think splitting up the builds resolves this too. But now I'm thinking it makes sense to document what can be done right now (one package version per build, based on the "root" SDK artifact) and come back to revise if splitting becomes possible.

Yes, all packages could use the same version (incl runtime, targeting packs), this would be the tagged version of source-build, which is the sdk version.

The main value of different versions(/splitting) is that users don't have to download some of the packages over in case of an upgrade (because the content hasn't changed). I think the users don't care much about this.

Reply:
@leecow @dagood shall we use the same version for all packages (that is: sdk version)?

Reply:
Having different package versions for components that are built together adds complexity for the package maintainer.

For example, `netstandard-targeting-pack-2.1` for `3.0.100` sdk will have the version `2.1.0` in the filesystem layout. Probably the `3.0.101` sdk will also have `2.1.0` in the filesystem.

If we're using the `2.1.0` as the package version, what should the maintainer do when `3.0.101` builds a package with the same version?

If we're using the sdk version, because there is a higher version number from the sdk (`3.0.100` to `3.0.101`), he can just push a new package.

Reply:
> Note that the Microsoft build does separate the Runtime vs. SDK build

@dagood, I guess `Microsoft.NETCore.App.Ref`, `Microsoft.NETCore.App.Host.<rid>` are built with the .NET Core runtime?
`Microsoft.AspNetCore.App.Ref` is built with ASP.NET Core shared framework?
`templates` are built with the SDK?
What builds `NETStandard.Library.Ref`, `Microsoft.WindowsDesktop.App.Ref`?

Reply:
I've pushed a commit that sets the package version for:
`dotnet-apphost-pack-[major].[minor]`, `dotnet-targeting-pack-[major].[minor]` to \<runtime version>.
`aspnetcore-targeting-pack-[major].[minor]` to \<aspnet core runtime version>.
`netstandard-targeting-pack-[major].[minor]` and `dotnet-templates-[major].[minor]` to \<sdk version>.